src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -283,8 +285,12 @@ async Task<MsalAccessTokenCacheItem> ITokenCacheInternal.FindAccessTokenAsync(Au
                     MsalErrorMessage.MultipleTokensMatched);
             }
 
+            msalAccessTokenCacheItem = FilterATByKeyId(msalAccessTokenCacheItem, requestParams);
```

Comment:
PoP tokens are bound to a kid, so if a different public / private key pair is used - do not return the token.

Reply:
what if there are multiple tokens bound to different KIDs? Possible?
Code will throw exception before filter is applied.

Reply:
It is possible to have multiple ATs bound to the same key, for example if they are for different client IDs or for different enviroments or different scopes.

But at this point in the `FindAccessToken` logic it's guaranteed that there are 0 or 1 tokens found in the cache. If more are found, an exception is thrown (see line 283).

Reply:
In this case, the access tokens will override each other. I chose this strategy because with PoP the key should be stable, i.e. one key per machine. The key can be lost, => new one will be generated, so we need to check AT is bound to same key. 

A different strategy would be to add `kid` to the access token cache key. This is considerably more complex and will break cache schema, i.e. old MSALs running side by side with new MSALs will fail. 

Third idea is to store PoP tokens separately from Bearer tokens. PoP tokens cache key would include `kid`. Adds quite a bit of complexity, but also flexibility. 

What do you think? Is it worth maybe discussing around this during spec review tomorrow?